resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/DriverAddressConfigurationStep.scala

```diff
@@ -0,0 +1,102 @@
+    driverLabels: Map[String, String],
+    submissionSparkConf: SparkConf,
+    clock: Clock) extends DriverConfigurationStep with Logging {
+  import DriverAddressConfigurationStep._
+
+  override def configureDriver(driverSpec: KubernetesDriverSpec): KubernetesDriverSpec = {
+    require(submissionSparkConf.getOption(DRIVER_BIND_ADDRESS_KEY).isEmpty,
+        s"$DRIVER_BIND_ADDRESS_KEY is not supported in Kubernetes mode, as the driver's hostname" +
+            s" will be managed via a Kubernetes service.")
+    require(submissionSparkConf.getOption(DRIVER_HOST_KEY).isEmpty,
+        s"$DRIVER_HOST_KEY is not supported in Kubernetes mode, as the driver's hostname will be" +
+            s" managed via a Kubernetes service.")
+
+    val preferredServiceName = s"$kubernetesResourceNamePrefix$DRIVER_SVC_POSTFIX"
+    val resolvedServiceName = if (preferredServiceName.length <= MAX_SERVICE_NAME_LENGTH) {
+      preferredServiceName
+    } else {
+      val randomServiceId = clock.getTimeMillis()
+      val shorterServiceName = s"spark-$randomServiceId$DRIVER_SVC_POSTFIX"
+      logWarning(s"Driver's hostname would preferably be $preferredServiceName, but this is" +
+          s" too long (must be <= 63 characters). Falling back to use $shorterServiceName" +
+              s" as the driver service's name.")
+      shorterServiceName
+    }
+
+    val driverPort = submissionSparkConf.getInt("spark.driver.port", DEFAULT_DRIVER_PORT)
+    val driverBlockManagerPort = submissionSparkConf.getInt(
+        org.apache.spark.internal.config.DRIVER_BLOCK_MANAGER_PORT.key, DEFAULT_BLOCKMANAGER_PORT)
+    val driverService = new ServiceBuilder()
+      .withNewMetadata()
+        .withName(resolvedServiceName)
+        .endMetadata()
+      .withNewSpec()
+        .withClusterIP("None")
+        .withSelector(driverLabels.asJava)
+        .addNewPort()
+          .withName(DRIVER_PORT_NAME)
+          .withPort(driverPort)
+          .withNewTargetPort(driverPort)
+          .endPort()
+        .addNewPort()
+          .withName(BLOCK_MANAGER_PORT_NAME)
+          .withPort(driverBlockManagerPort)
+          .withNewTargetPort(driverBlockManagerPort)
+          .endPort()
+        .endSpec()
+      .build()
+
+    val namespace = submissionSparkConf.get(KUBERNETES_NAMESPACE)
+    val driverHostname = s"${driverService.getMetadata.getName}.$namespace.svc.cluster.local"
+    val resolvedSparkConf = driverSpec.driverSparkConf.clone()
+        .set(org.apache.spark.internal.config.DRIVER_BIND_ADDRESS, driverHostname)
+        .set(org.apache.spark.internal.config.DRIVER_HOST_ADDRESS, driverHostname)
+        .set("spark.driver.port", driverPort.toString)
+        .set(
+            org.apache.spark.internal.config.DRIVER_BLOCK_MANAGER_PORT, driverBlockManagerPort)
+
+    driverSpec.copy(
+        driverSparkConf = resolvedSparkConf,
+        otherKubernetesResources = driverSpec.otherKubernetesResources ++ Seq(driverService))
```

Comment:
My understanding is those dependent Kubernetes resources of the driver gets created after the driver Pod gets created. If this is correct, then this is problematic as there may not be an Endpoint object created for the driver Pod because the Pod exists before the service is in place.